In [3]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

* **Load the dataset from “sampleData.csv”.  Create training and test datasets using random splitting (70-30)%. Print the train and test data. Create a linear regression model and report the model performance. (The last feature is the target variable)**

In [4]:
#Load data
df = pd.read_csv('sample_dataset.csv')

In [5]:
df.head()

Subject ID  sample ID  feature_0  feature_1  feature_2  feature_3  \
0           1          1   0.049592   0.545805   0.771077   0.387171   
1           1          2   0.784486   0.361281   0.681752   0.494981   
2           1          3   0.017367   0.661350   0.849570   0.833185   
3           1          4   0.197825   0.227258   0.177861   0.179123   
4           1          5   0.899094   0.817947   0.889295   0.205932   

   feature_4  feature_5  feature_6  feature_7  ...  feature_22  feature_23  \
0   0.618290   0.869132   0.139296   0.454377  ...    0.118728    0.971647   
1   0.490758   0.132277   0.475217   0.041860  ...    0.836944    0.694952   
2   0.591147   0.357999   0.583466   0.681106  ...    0.006386    0.937494   
3   0.397492   0.010650   0.111507   0.719310  ...    0.177616    0.517563   
4   0.005874   0.687918   0.874622   0.523576  ...    0.921892    0.784476   

   feature_24  feature_25  feature_26  feature_27  feature_28  feature_29  \
0    0.517703    0.390375    0.364423    0.948268    0.921592    0.986265   
1    0.990620    0.573024    0.627210    0.670506    0.339624    0.238391   
2    0.343176    0.816586    0.295563    0.078583    0.724032    0.847065   
3    0.336766    0.315501    0.478920    0.203019    0.190466    0.739788   
4    0.778715    0.517458    0.289470    0.185447    0.960247    0.746802   

   feature_30  feature_31  
0    0.983804    0.892498  
1    0.141122    0.989782  
2    0.704773    0.793935  
3    0.210447    0.084941  
4    0.891682    0.733321  

[5 rows x 34 columns]

In [6]:
#get the size of the dataset
df.shape

(150, 34)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Subject ID  150 non-null    int64  
 1   sample ID   150 non-null    int64  
 2   feature_0   150 non-null    float64
 3   feature_1   150 non-null    float64
 4   feature_2   150 non-null    float64
 5   feature_3   150 non-null    float64
 6   feature_4   150 non-null    float64
 7   feature_5   150 non-null    float64
 8   feature_6   150 non-null    float64
 9   feature_7   150 non-null    float64
 10  feature_8   150 non-null    float64
 11  feature_9   150 non-null    float64
 12  feature_10  150 non-null    float64
 13  feature_11  150 non-null    float64
 14  feature_12  150 non-null    float64
 15  feature_13  150 non-null    float64
 16  feature_14  150 non-null    float64
 17  feature_15  150 non-null    float64
 18  feature_16  150 non-null    float64
 19  feature_17  150 non-null    f

In [8]:
df.describe()

Subject ID   sample ID   feature_0   feature_1   feature_2   feature_3  \
count  150.000000  150.000000  150.000000  150.000000  150.000000  150.000000   
mean    15.500000    3.000000    0.519458    0.515589    0.551360    0.518398   
std      8.684438    1.418951    0.293991    0.285416    0.270573    0.281292   
min      1.000000    1.000000    0.007503    0.001806    0.010256    0.013015   
25%      8.000000    2.000000    0.238741    0.268906    0.345227    0.269774   
50%     15.500000    3.000000    0.580427    0.519306    0.594042    0.513395   
75%     23.000000    4.000000    0.770542    0.769700    0.787061    0.750345   
max     30.000000    5.000000    0.998945    0.996198    0.994709    0.996611   

        feature_4   feature_5   feature_6   feature_7  ...  feature_22  \
count  150.000000  150.000000  150.000000  150.000000  ...  150.000000   
mean     0.507698    0.477344    0.478999    0.501294  ...    0.520925   
std      0.285513    0.282899    0.281569    0.270244  ...    0.300841   
min      0.005874    0.004173    0.000843    0.009966  ...    0.002228   
25%      0.261920    0.256708    0.269379    0.282559  ...    0.259045   
50%      0.524667    0.475332    0.454109    0.516709  ...    0.522819   
75%      0.761900    0.686078    0.723747    0.720259  ...    0.826411   
max      0.999945    0.991820    0.989347    0.973196  ...    0.999487   

       feature_23  feature_24  feature_25  feature_26  feature_27  feature_28  \
count  150.000000  150.000000  150.000000  150.000000  150.000000  150.000000   
mean     0.541186    0.507834    0.512649    0.475324    0.534528    0.494043   
std      0.284148    0.293735    0.287857    0.306199    0.300744    0.287329   
min      0.005880    0.004842    0.005909    0.001088    0.000308    0.008065   
25%      0.296334    0.263791    0.256659    0.194541    0.288860    0.259479   
50%      0.544364    0.496645    0.534315    0.467230    0.577002    0.476170   
75%      0.786015    0.768850    0.755917    0.765318    0.804632    0.728498   
max      0.999388    0.990620    0.995900    0.994541    0.998447    0.990492   

       feature_29  feature_30  feature_31  
count  150.000000  150.000000  150.000000  
mean     0.553169    0.479062    0.495994  
std      0.277459    0.300696    0.309316  
min      0.006614    0.000409    0.004522  
25%      0.326223    0.190981    0.222037  
50%      0.538039    0.510464    0.495930  
75%      0.805034    0.706531    0.747992  
max      0.992272    0.999007    0.989782  

[8 rows x 34 columns]

In [9]:
df.isnull().sum()

Subject ID    0
sample ID     0
feature_0     0
feature_1     0
feature_2     0
feature_3     0
feature_4     0
feature_5     0
feature_6     0
feature_7     0
feature_8     0
feature_9     0
feature_10    0
feature_11    0
feature_12    0
feature_13    0
feature_14    0
feature_15    0
feature_16    0
feature_17    0
feature_18    0
feature_19    0
feature_20    0
feature_21    0
feature_22    0
feature_23    0
feature_24    0
feature_25    0
feature_26    0
feature_27    0
feature_28    0
feature_29    0
feature_30    0
feature_31    0
dtype: int64

In [10]:
#Get the features and labels
X = df.iloc[:,:-1]
y = df.iloc[:, -1]

In [11]:
from sklearn.model_selection import train_test_split

In [12]:
#Split train and test dataset 7:3
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [13]:
#Print the train and test set
X_train.head()

Subject ID  sample ID  feature_0  feature_1  feature_2  feature_3  \
81           17          2   0.965936   0.878103   0.386270   0.218148   
133          27          4   0.378926   0.822470   0.810673   0.796805   
137          28          3   0.602190   0.836786   0.739193   0.074982   
75           16          1   0.647087   0.583075   0.240903   0.686149   
109          22          5   0.299556   0.944794   0.245175   0.153634   

     feature_4  feature_5  feature_6  feature_7  ...  feature_21  feature_22  \
81    0.999945   0.987537   0.633259   0.067809  ...    0.527833    0.671222   
133   0.799147   0.493268   0.473827   0.855313  ...    0.687249    0.297938   
137   0.321208   0.352287   0.963184   0.189590  ...    0.794904    0.617638   
75    0.032975   0.141510   0.174590   0.020943  ...    0.030765    0.441087   
109   0.134723   0.196001   0.702827   0.628675  ...    0.629802    0.251942   

     feature_23  feature_24  feature_25  feature_26  feature_27  feature_28  \
81     0.552047    0.668078    0.439723    0.485720    0.124052    0.667793   
133    0.288673    0.078337    0.762598    0.206676    0.013747    0.909569   
137    0.762673    0.614363    0.018864    0.269345    0.190072    0.258485   
75     0.632321    0.046242    0.282524    0.895349    0.043557    0.970534   
109    0.594300    0.306838    0.766697    0.016570    0.510205    0.540352   

     feature_29  feature_30  
81     0.992199    0.157581  
133    0.192709    0.049537  
137    0.169470    0.870730  
75     0.267381    0.010865  
109    0.267505    0.856083  

[5 rows x 33 columns]

In [14]:
y_train.head()

81     0.664231
133    0.696368
137    0.242381
75     0.628457
109    0.970861
Name: feature_31, dtype: float64

In [15]:
X_test.head()

Subject ID  sample ID  feature_0  feature_1  feature_2  feature_3  \
73           15          4   0.948736   0.701764   0.758984   0.764294   
18            4          4   0.153687   0.691285   0.620214   0.268056   
118          24          4   0.905840   0.408101   0.541389   0.341156   
78           16          4   0.587969   0.612320   0.688419   0.298370   
76           16          2   0.085164   0.364504   0.698854   0.740021   

     feature_4  feature_5  feature_6  feature_7  ...  feature_21  feature_22  \
73    0.302214   0.316796   0.133772   0.972011  ...    0.558057    0.373930   
18    0.730510   0.441093   0.817469   0.592335  ...    0.118995    0.471973   
118   0.297925   0.484835   0.182228   0.225059  ...    0.227989    0.397496   
78    0.528624   0.252250   0.110105   0.860619  ...    0.777804    0.999487   
76    0.512721   0.834822   0.712927   0.849809  ...    0.021494    0.357855   

     feature_23  feature_24  feature_25  feature_26  feature_27  feature_28  \
73     0.760234    0.495353    0.619501    0.815067    0.746257    0.952560   
18     0.783188    0.916690    0.991552    0.932083    0.032833    0.561273   
118    0.645922    0.379403    0.461324    0.834726    0.500437    0.058176   
78     0.942200    0.717267    0.927554    0.830067    0.188607    0.176468   
76     0.449751    0.100079    0.835015    0.485891    0.589269    0.357913   

     feature_29  feature_30  
73     0.765208    0.001485  
18     0.446163    0.150041  
118    0.755119    0.033973  
78     0.358316    0.416079  
76     0.802182    0.160600  

[5 rows x 33 columns]

In [16]:
y_test.head()

73     0.036916
18     0.515118
118    0.748968
78     0.903448
76     0.917985
Name: feature_31, dtype: float64

In [17]:
#Build linear model
from sklearn.linear_model import LinearRegression

In [18]:
lr = LinearRegression()

In [19]:
lr.fit(X_train, y_train)

LinearRegression()

In [20]:
pred = lr.predict(X_test)

In [21]:
#Validate the model
from sklearn.metrics import mean_squared_error, mean_absolute_error, confusion_matrix, classification_report

In [22]:
print('Mean Square Error: ', round(mean_squared_error(y_test, pred),2))
print('Mean Absolute Error: ', round(mean_absolute_error(y_test, pred),2))
print('Classification report: ', classification_report(y_test.round(), pred.round()))

Mean Square Error:  0.12
Mean Absolute Error:  0.3
Classification report:                precision    recall  f1-score   support

         0.0       0.50      0.57      0.53        21
         1.0       0.57      0.50      0.53        24

    accuracy                           0.53        45
   macro avg       0.54      0.54      0.53        45
weighted avg       0.54      0.53      0.53        45



* **Create training and test dataset using leave-one-out based on Subject ID. Print the train and test dataset for the first iteration only.  Create a linear regression model and compare the performance with Q-1. Explain your findings.**

In [28]:
from sklearn.model_selection import cross_val_score, LeaveOneOut, LeaveOneGroupOut

In [29]:
X = df.iloc[:,:-1].values
y = df.iloc[:, -1].values

In [30]:
logo = LeaveOneGroupOut()

In [32]:
groups = df['Subject ID']

y_true, y_pred = list(), list()

for i, (train_index, test_index) in enumerate(logo.split(X, y, groups)):
    
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]
    
    if i == 1:
        print('X_train: ', X_train)
        print('X_test: ', X_test)
        print('y_train: ', y_train)
        print('y_test: ', y_test)
        
    lr_model = LinearRegression()
    lr_model.fit(X_train, y_train)
    
    predict = lr_model.predict(X_test)
    
    y_true.append(y_test[0])
    y_pred.append(predict[0])
    
print('Mean Square Error: ', round(mean_squared_error(y_true, y_pred),2))
print('Mean Absolute Error: ', round(mean_absolute_error(y_true, y_pred),2))
print('Classification report: ', classification_report(np.around(np.array(y_true)), np.around(np.array(y_pred))))    

X_train:  [[1.00000000e+00 1.00000000e+00 4.95920928e-02 ... 9.21592123e-01
  9.86265245e-01 9.83803733e-01]
 [1.00000000e+00 2.00000000e+00 7.84486332e-01 ... 3.39623796e-01
  2.38391490e-01 1.41121801e-01]
 [1.00000000e+00 3.00000000e+00 1.73669900e-02 ... 7.24031975e-01
  8.47064837e-01 7.04772812e-01]
 ...
 [3.00000000e+01 3.00000000e+00 7.31092685e-01 ... 6.52748112e-01
  2.60084704e-01 4.19321041e-01]
 [3.00000000e+01 4.00000000e+00 2.83677687e-01 ... 2.25610011e-02
  8.07065153e-01 7.95217481e-01]
 [3.00000000e+01 5.00000000e+00 7.00753702e-01 ... 7.28635928e-01
  9.92271704e-01 5.05454738e-01]]
X_test:  [[2.         1.         0.16878321 0.42827873 0.54215273 0.45705544
  0.38854539 0.54591027 0.95060005 0.0627298  0.68175111 0.59199996
  0.09369669 0.19995852 0.56439894 0.43724959 0.74318467 0.36635595
  0.77789614 0.281484   0.75807968 0.18023986 0.01097322 0.19509284
  0.90835907 0.94590834 0.84126831 0.46272227 0.19416378 0.85619543
  0.39953687 0.80800952 0.83870045]
 [2. 

As we can see that all the metrics from the train test split and the leave one group out method are pretty the same: mean square error is around 0.12, mean absolute error is 0.3, accuracy is around 50%.